File: app/src/main/java/org/lenchan139/ncbookmark/v2/AddBookmarkActivityV2.java
package org.lenchan139.ncbookmark.v2;

import android.content.DialogInterface;
import android.os.AsyncTask;
import android.os.Bundle;
import android.support.design.widget.FloatingActionButton;
import android.support.design.widget.Snackbar;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.util.Base64;
import android.util.Log;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Toast;

import org.json.JSONArray;
import org.json.JSONException;
import org.jsoup.Connection;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.lenchan139.ncbookmark.R;

import java.io.IOException;
import java.net.URL;
import java.util.ArrayList;

public class AddBookmarkActivityV2 extends AppCompatActivity {

    String urlNt,username,password;
    String[] tags;
    String login;
    EditText edtUrl,edtTitle,edtDescr,edtTag;
    Button btnSelectTag,btnSubmit;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_add_bookmark_v2);
        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        edtDescr = (EditText) findViewById(R.id.description);
        edtTag = (EditText) findViewById(R.id.tag);
        edtTitle = (EditText) findViewById(R.id.title);
        edtUrl = (EditText) findViewById(R.id.url);
        btnSelectTag = (Button) findViewById(R.id.selectTag);
        btnSubmit = (Button) findViewById(R.id.submit);

        Window window = getWindow();
        window.addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS);
        window.setStatusBarColor(getResources().getColor(R.color.colorPrimaryDark));
        FloatingActionButton fab = (FloatingActionButton) findViewById(R.id.fab);
        fab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Snackbar.make(view, "Replace with your own action", Snackbar.LENGTH_LONG)
                        .setAction("Action", null).show();
            }
        });
        fab.setVisibility(View.GONE);
        //
         urlNt = getIntent().getStringExtra("url");
         username = getIntent().getStringExtra("username");
         password = getIntent().getStringExtra("password");
        login = username + ":" + password;
        if(urlNt != null && username != null && password !=null){
            btnSubmit.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    new AddBookmarkTask().execute();
                }
            });
        }else{

        }

        btnSelectTag.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                new fetchTagsTask().execute();
            }
        });
    }

    private class fetchTagsTask extends AsyncTask<URL, Integer, Long> {
        String urlSe = "/index.php/apps/bookmarks/public/rest/v2/tag";
        final String base64login = new String(Base64.encode(login.getBytes(),0));
        Document result;

        @Override
        protected Long doInBackground(URL... params) {

            try {
                result = Jsoup.connect(urlNt + urlSe)
                        .ignoreContentType(true)
                        .header("Authorization", "Basic " + base64login)
                        .method(Connection.Method.GET)
                        .execute().parse();
            } catch (IOException e) {
                e.printStackTrace();
            }
            return null;
        }

        @Override
        protected void onPostExecute(Long aLong) {
            Log.v("testAllTags",result.body().text());
            try {
                JSONArray jsonArray = new JSONArray(result.body().text());
                tags = new String[jsonArray.length()];
                for (int i=0;i<jsonArray.length();i++){
                    tags[i] = jsonArray.get(i).toString();
                    Log.v("testTagConverting",jsonArray.get(i).toString());
                }

            } catch (JSONException e) {
                e.printStackTrace();
                tags = null;
            }
            if(tags != null){
                AlertDialog.Builder dialog = new AlertDialog.Builder(AddBookmarkActivityV2.this);
                dialog.setTitle("Exciting Tag");
                dialog.setItems(tags, new DialogInterface.OnClickListener(){
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        edtTag.setText(tags[which]);
                    }
                });
                dialog.setNegativeButton("Cancel",null);
                dialog.show();
            }
            super.onPostExecute(aLong);
        }

    }

    private class AddBookmarkTask extends AsyncTask<URL, Integer, Long> {
        String tag,url,title,des;
        Document result;
        boolean no_error = true;
        String error_msg;
        String urlSe = "/index.php/apps/bookmarks/public/rest/v2/bookmark";
        final String base64login = new String(Base64.encode(login.getBytes(),0));
        @Override
        protected void onPreExecute() {
            tag = edtTag.getText().toString();
            url = edtUrl.getText().toString();
            title = edtTitle.getText().toString();
            des = edtDescr.getText().toString();
            if(tag == null || url == null || title == null){
                no_error = false;
                Toast.makeText(AddBookmarkActivityV2.this, "Please tag, url and title cannot be empty.", Toast.LENGTH_SHORT).show();
            }
            if(des == null){
                des = "";
            }
            super.onPreExecute();
        }

        @Override
        protected Long doInBackground(URL... params) {
            if(no_error){

                try {
                    result = Jsoup.connect(urlNt + urlSe)
                            .ignoreContentType(true)
                            .header("Authorization", "Basic " + base64login)
                            .method(Connection.Method.POST)
                            .data("url",url)
                            .data("item[tags][]",tag)
                            .data("title",title)
                            .data("description",des)
                            .execute().parse();
                } catch (IOException e) {
                    e.printStackTrace();
                    error_msg = e.getMessage();
                }
            }
            return null;
        }

        @Override
        protected void onPostExecute(Long aLong) {
            if(error_msg != null){
                Toast.makeText(AddBookmarkActivityV2.this, error_msg, Toast.LENGTH_SHORT).show();
            }else{
                Toast.makeText(AddBookmarkActivityV2.this, "Bookmark Added!", Toast.LENGTH_SHORT).show();
                AddBookmarkActivityV2.this.finish();
            }
            super.onPostExecute(aLong);
        }
    }
}

File: app/src/main/java/org/lenchan139/ncbookmark/v2/AddBookmarkActivityV2.kt
package org.lenchan139.ncbookmark.v2

import android.os.AsyncTask
import android.os.Bundle
import android.support.design.widget.FloatingActionButton
import android.support.design.widget.Snackbar
import android.support.v7.app.AlertDialog
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.Toolbar
import android.util.Base64
import android.util.Log
import android.view.View
import android.view.WindowManager
import android.widget.Button
import android.widget.EditText
import android.widget.Toast

import org.json.JSONArray
import org.json.JSONException
import org.jsoup.Connection
import org.jsoup.Jsoup
import org.jsoup.nodes.Document
import org.lenchan139.ncbookmark.Class.TagsItem
import org.lenchan139.ncbookmark.Constants
import org.lenchan139.ncbookmark.R

import java.io.IOException
import java.net.URL

class AddBookmarkActivityV2 : AppCompatActivity() {

    internal var urlNt: String? = null
    internal var username: String? = null
    internal var password: String? = null
    internal lateinit var tags: Array<String?>
    internal lateinit var login: String
    internal lateinit var edtUrl: EditText
    internal lateinit var edtTitle: EditText
    internal lateinit var edtDescr: EditText
    internal lateinit var edtTag: EditText
    internal lateinit var btnSelectTag: Button
    internal lateinit var btnSubmit: Button
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_add_bookmark_v2)
        val toolbar = findViewById(R.id.toolbar) as Toolbar
        setSupportActionBar(toolbar)
        edtDescr = findViewById(R.id.description) as EditText
        edtTag = findViewById(R.id.tag) as EditText
        edtTitle = findViewById(R.id.title) as EditText
        edtUrl = findViewById(R.id.url) as EditText
        btnSelectTag = findViewById(R.id.selectTag) as Button
        btnSubmit = findViewById(R.id.submit) as Button

        val window = window
        window.addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS)
        window.statusBarColor = resources.getColor(R.color.colorPrimaryDark)
        val fab = findViewById(R.id.fab) as FloatingActionButton
        fab.setOnClickListener { view ->
            Snackbar.make(view, "Replace with your own action", Snackbar.LENGTH_LONG)
                    .setAction("Action", null).show()
        }
        fab.visibility = View.GONE
        //for out call use
        var inUrl = intent.getStringExtra("inUrl")
        var inTitle = intent.getStringExtra("inTitle")
        var inTag = intent.getStringExtra("inTag")
        //
        var sp = getSharedPreferences("data", 0)
        urlNt = sp.getString("url",null)
        username = sp.getString("username",null)
        password = sp.getString("password",null)
        var apiVer = sp.getInt("apiVersion",0)
        login = username + ":" + password
        if (urlNt != null && username != null && password != null && apiVer >= 2) {
            btnSubmit.setOnClickListener { AddBookmarkTask().execute() }
        } else {
            Toast.makeText(this,"maybe you are not login yet.",Toast.LENGTH_SHORT).show()
            finish()
        }

        if(inUrl!=null){
            edtUrl.setText(inUrl)
        }
        if(inTag!=null){
            edtTag.setText(inTag)
        }
        if(inTitle!=null){
            edtTitle.setText(inTitle)
        }



        btnSelectTag.setOnClickListener { fetchTagsTask().execute() }
    }

    private inner class fetchTagsTask : AsyncTask<URL, Int, Long>() {
        internal var urlSe = Constants.V2_API_ENDPOINT + "tag"
        internal val base64login = String(Base64.encode(login.toByteArray(), 0))
        internal lateinit var result: Document

        override fun doInBackground(vararg params: URL): Long? {

            try {
                result = Jsoup.connect(urlNt!! + urlSe)
                        .ignoreContentType(true)
                        .header("Authorization", "Basic " + base64login)
                        .method(Connection.Method.GET)
                        .execute().parse()
            } catch (e: IOException) {
                e.printStackTrace()
            }

            return null
        }
        override fun onPostExecute(aLong: Long?) {
            Log.v("testAllTags", result.body().text())
            var err_check = true
            try {
                val jsonArray = JSONArray(result.body().text())
                tags = arrayOfNulls<String>(jsonArray.length())

                for (i in 0..jsonArray.length() - 1) {
                    tags[i] = jsonArray.get(i).toString()
                    Log.v("testTagConverting", jsonArray.get(i).toString())
                }

            } catch (e: JSONException) {
                e.printStackTrace()
                err_check = false
            }

            if (err_check) {
                val dialog = AlertDialog.Builder(this@AddBookmarkActivityV2)
                dialog.setTitle("Exciting Tag")
                dialog.setItems(tags) { dialog, which -> edtTag.setText(tags!![which]) }
                dialog.setNegativeButton("Cancel", null)
                dialog.show()
            }
            super.onPostExecute(aLong)
        }

    }

    private inner class AddBookmarkTask : AsyncTask<URL, Int, Long>() {
        internal lateinit var tag: Array<String>
        internal var url: String? = null
        internal var title: String? = null
        internal var des: String? = null
        internal lateinit var result: Document
        internal var no_error = true
        internal var error_msg: String? = null
        internal var urlSe = Constants.V2_API_ENDPOINT + "bookmark"
        internal val base64login = String(Base64.encode(login.toByteArray(), 0))

        override fun onPreExecute() {
            val tampTag = TagsItem().strToArray(edtTag.text.toString())
            if(tampTag!= null){
                tag = tampTag
            }
            url = edtUrl.text.toString()
            title = edtTitle.text.toString()
            des = edtDescr.text.toString()
            if ( url == null || title == null) {
                no_error = false
                Toast.makeText(this@AddBookmarkActivityV2, "Please tag, url and title cannot be empty.", Toast.LENGTH_SHORT).show()
            }
            if (des == null) {
                des = ""
            }
            super.onPreExecute()
        }

        override fun doInBackground(vararg params: URL): Long? {
            if (no_error) {

                try {
                    var jsoup = Jsoup.connect(urlNt!! + urlSe)
                            .ignoreContentType(true)
                            .header("Authorization", "Basic " + base64login)
                            .method(Connection.Method.POST)
                            .data("url", url)
                            .data("title", title)
                            .data("description", des)
                           for(i in 0..tag.size-1){
                               jsoup.data("item[tags][]", tag.get(i).trim())
                           }
                    result = jsoup.execute().parse()

                } catch (e: IOException) {
                    e.printStackTrace()
                    error_msg = e.message
                }

            }
            return null
        }

        override fun onPostExecute(aLong: Long?) {
            if (error_msg != null) {
                Toast.makeText(this@AddBookmarkActivityV2, error_msg, Toast.LENGTH_SHORT).show()
            } else {
                Toast.makeText(this@AddBookmarkActivityV2, "Bookmark Added!", Toast.LENGTH_SHORT).show()
                this@AddBookmarkActivityV2.finish()
            }
            super.onPostExecute(aLong)
        }
    }
}
